src/Microsoft.AspNetCore.Hosting/WebHostExtensions.cs

```diff
@@ -16,18 +20,33 @@ public static class WebHostExtensions
         /// <param name="host">The <see cref="IWebHost"/> to run.</param>
         public static void Run(this IWebHost host)
         {
+            var done = new ManualResetEventSlim(false);
             using (var cts = new CancellationTokenSource())
             {
-                Console.CancelKeyPress += (sender, eventArgs) =>
+                Action shutdown = () =>
                 {
-                    Console.WriteLine("Application is shutting down...");
-                    cts.Cancel();
+                    if (!cts.IsCancellationRequested)
+                    {
+                        Console.WriteLine("Application is shutting down...");
+                        cts.Cancel();
+                    }
 
+                    done.Wait();
+                };
+
+#if NETSTANDARD1_5
+                var assemblyLoadContext = AssemblyLoadContext.GetLoadContext(typeof(WebHostExtensions).GetTypeInfo().Assembly);
+                assemblyLoadContext.Unloading += context => shutdown();
+#endif
+                Console.CancelKeyPress += (sender, eventArgs) =>
+                {
+                    shutdown();
```

Comment:
The behavior of shutdown() is now synchronous vs. asynchronous. So long as CancelKeyPress comes in on a random background thread, this shouldn't matter. Just pointing out there may be a behavioral difference from the existing code.